### DISCUSSION TITLE
How to perform data validations programatically?

### DISCUSSION THREAD

**Discussion by joanlopez** [cuelang/cue#810] [Q&A] [answered]
Hi folks! 

I'm trying to add support for Cue validations into another Go tool, so basically I'm wondering how I can perform data validations programatically. 

Let's say in a similar way I'd run the `go vet` command with a data file (JSON, YAML, etc) and a CUE file with some data constrains & schema.

How can I achieve something similar programatically, considering, for example, I have the bytes of each file in a variable? 

Thanks!! 🙏🏻

**Comment by mpvl** [comment, marked as answer]
There is an extensive API for this. See `cuelang.org/go/cue`, `cuelang.org/go/cue/load` and friends.

Validation is typically done by getting two `cue.Value`s and unify them using `cue.Unify`. The `Validate` method allows you to be more specific about which errors to get.

You can also validate Go values directly. See `cuelang.org/go/encoding/gocode`.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/810.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
